tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs

```diff
@@ -0,0 +1,280 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading.Tasks;
+using DevHome.PI.Helpers;
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+using Microsoft.UI.Xaml.Media.Imaging;
+using Windows.Devices.Geolocation;
+
+namespace DevHome.PI.Controls;
+
+internal sealed class ExternalToolsManagementButton : Button
+{
+    private const string _UnregisterButtonText = "\uECC9";
+
+    private readonly Microsoft.UI.Dispatching.DispatcherQueue _dispatcher;
+    private readonly string _pinMenuItemText = CommonHelper.GetLocalizedString("PinMenuItemText");
+    private readonly string _unpinMenuItemText = CommonHelper.GetLocalizedString("UnpinMenuItemRawText");
+    private readonly string _unregisterMenuItemText = CommonHelper.GetLocalizedString("UnregisterMenuItemRawText");
+
+    public event EventHandler<ExternalTool>? ExternalToolLaunchRequest;
+
+    public ExternalToolsManagementButton()
+    {
+        _dispatcher = Microsoft.UI.Dispatching.DispatcherQueue.GetForCurrentThread();
+
+        InitializeExternalTools();
+    }
+
+    private void InitializeExternalTools()
+    {
+        if (ExternalToolsHelper.Instance.AllExternalTools.Count > 0)
+        {
+            var flyout = new MenuFlyout();
+            this.ContextFlyout = flyout;
+
+            foreach (var tool in ExternalToolsHelper.Instance.AllExternalTools)
+            {
+                AddExternalToolToContextMenu(flyout, tool);
+                tool.PropertyChanged += ExternalToolItem_PropertyChanged;
+            }
+        }
+
+        // We have to cast to INotifyCollectionChanged explicitly because the CollectionChanged
+        // event in ReadOnlyObservableCollection is protected.
+        ((INotifyCollectionChanged)ExternalToolsHelper.Instance.AllExternalTools).CollectionChanged += ExternalTools_CollectionChanged;
+    }
+
+    private void ExternalTools_CollectionChanged(object? sender, NotifyCollectionChangedEventArgs e)
+    {
+        _dispatcher.TryEnqueue(() =>
+        {
+            if (e.Action == NotifyCollectionChangedAction.Add && e.NewItems is not null)
+            {
+                foreach (ExternalTool tool in e.NewItems)
+                {
+                    // If this button doens't have a flyout, create one
+                    if (this.ContextFlyout is not MenuFlyout flyout)
```

Comment:
I don't understand why ContextFlyout isn't a MenuFlyout is a signal that the button does not have a flyout.

Reply:
Could have done

if (this.ContextFlyout is null)
...

MenuFlyout flyout = this.ContextFlyout as MenuFlyout;
Debug.Assert(flyout is not null);

But this seemed more compact and accomplished what I was looking for.

Reply:
What if `ContextMenu` is not null and not a `MenuFlyout`?  Also, if I am confused about the meaning of the comment, other people will be as well.  Consider adding more description explaining why something that isn't a `MenuFlyout` equates to "The button does not have a flyout"

Reply:
If that's the case, then bad things have happened, meaning that somebody else is mucking with this control's flyout. I could add an assert that nobody else is mucking with my context menu...